fix(service-worker): clear service worker cache in safety worker

naveedahmed1 opened the pull request:
clear angular service worker cache in safety worker to ensure stale
or broken contents are not served in future requests

Fixes #43163

## PR Checklist
Please check if your PR fulfills the following requirements:

- [ x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit

## PR Type
What kind of change does this PR introduce?

- [ x] Bugfix
- [ x] Improvements in safety worker

## What is the current behavior?
Currently, safety worker only unregisters old service worker, while leaving the service worker cache. This creates problem when cache contains broken contents and sometimes causes new service worker to serve the contents from the old cache. Hence failing the purpose of safety worker.

Issue Number: 43163


## What is the new behavior?
After unregistering old service worker, clear cache of angular service worker. This ensure that the new service worker doesn't serve stale or broken contents.

## Does this PR introduce a breaking change?

- [ ] Yes
- [x ] No

gkalpak commented on packages/service-worker/safety-worker.js:
```diff
@@ -18,3 +18,16 @@ self.addEventListener('activate', event => {
   });
 });
+
+caches.keys().then(cacheKeys =>
+  Promise.all(
+    cacheKeys.map(cacheKey => {
+      const ngswRegex = /^(ngsw).*/;
+      if (ngswRegex.test(cacheKey)) {
+        return caches
+          .open(cacheKey)
+          .then(cache => cache.keys().then(requests => requests.map(req => cache.delete(req))));
+      }
+    })
+  )
+);
```
A few suggestions/improvements:

- It is safer to delete the caches inside the `activate` event handler (to avoid deleting the caches before this SW takes over).
- There is no need to delete each request in a cache. We can just delete the whole cache.
- We should be targeting caches that start with `ngsw:` (notice the double-colon at the end).
- We should ensure the SW instance remains active while we delete the caches (by passing a promise to `event.waitUntil()`.

Based on the above, the code look like this (unless I've missed something 😅):

```suggestion
event.waitUntil(caches.keys().then(cacheNames => {
  const ngswCacheNames = cacheNames.filter(name => /^ngsw:/.test(name));
  return Promise.all(ngswCacheNames.map(name => caches.delete(name)));
}));
```

gkalpak reviewed (changes requested):
Thx, @naveedahmed1 :+1:

I've left a couple of suggestions. Could you also mention the change in the [Safety Worker][1] section of [Service worker in production][2] guide? 🙏

[1]: https://github.com/angular/angular/blame/3d101f989dd47557614bc2c1a28cef8dc1792f83/aio/content/guide/service-worker-devops.md#L325
[2]: https://angular.io/guide/service-worker-devops

naveedahmed1 commented:
Thanks @gkalpak I have incorporated the changes you suggested. I hope everything is in order now :D

gkalpak commented on packages/service-worker/safety-worker.js:
```diff
@@ -18,3 +18,8 @@ self.addEventListener('activate', event => {
+event.waitUntil(caches.keys().then(cacheNames => {
+  const ngswCacheNames = cacheNames.filter(name => /^ngsw:/.test(name));
+  return Promise.all(ngswCacheNames.map(name => caches.delete(name)));
+}));
```
Thx for updating the code, @naveedahmed1 :+1:
Can you please move this block inside the `activate` event handler above?

```diff
 self.addEventListener('activate', event => {
   event.waitUntil(self.clients.claim());
   self.registration.unregister().then(() => {
     console.log('NGSW Safety Worker - unregistered old service worker');
   });
+  event.waitUntil(caches.keys().then(cacheNames => {
+    const ngswCacheNames = cacheNames.filter(name => /^ngsw:/.test(name));
+    return Promise.all(ngswCacheNames.map(name => caches.delete(name)));
+  }));
 });
```

(And while you're at it, you could also wrap the `self.registration.unregister().then(...)` call in `event.waitUntil()` :wink:)

naveedahmed1 replied:
@gkalpak I have updated the code, can you please take a look?

gkalpak commented on aio/content/guide/service-worker-devops.md:
```diff
@@ -326,8 +326,9 @@ essentially self-destructing.
 script `safety-worker.js`, which when loaded will unregister itself
-from the browser. This script can be used as a last resort to get rid
-of unwanted service workers already installed on client pages.
+from the browser and remove the service worker cache. This script can 
```
Let's use `caches`:

```suggestion
from the browser and remove the service worker caches. This script can 
```

gkalpak commented on aio/content/guide/service-worker-devops.md:
```diff
@@ -340,7 +341,7 @@ old Service Worker URL forever.
 
 This script can be used both to deactivate `@angular/service-worker`
 as well as any other Service Workers which might have been served in
-the past on your site.
+the past on your site along with the corresponding caches.
```
Let's move the mention about caches next to `@angular/service-worker` to avoid confusion (since the script only removes ngsw-specific caches - not any caches that might be used by other SW scripts):

```suggestion
This script can be used both to deactivate `@angular/service-worker`
(and remove the corresponding caches) as well as any other Service
Workers which might have been served in the past on your site.
```

naveedahmed1 replied:
@gkalpak I have made the changes and both safety worker and docs updates are not in one commit. Can you please take a look, I hope every thing is fine now. Thank you so much once again for your time and reviewing my PR.

gkalpak reviewed (changes requested):
Thx for making the changes, @naveedahmed1 :+1:
I have left a couple of small suggestions regarding the docs. Can you also please ensure that there is only a single commit on the PR (since this is conceptually a single change) :pray:

gkalpak reviewed (changes requested):
Thx, @naveedahmed1 :+1:

It looks good to me, except I still see 5 commits on the PR: https://github.com/angular/angular/pull/43324/commits
Can you please make sure there is only a single commit?

naveedahmed1 committed:
fix(service-worker): clear service worker cache in safety worker

clear angular service worker cache in safety worker to ensure stale
or broken contents are not served in future requests

Fixes #43163
